src/antsibull_core/dependency_files.py

```diff
@@ -164,12 +190,32 @@ def write(
 
 
 class BuildFile:
-    def __init__(self, build_file: str) -> None:
-        self.filename: str = build_file
+    def __init__(self, build_file: StrPath) -> None:
+        self.filename: str = str(build_file)
 
-    def parse(self) -> DependencyFileData:
-        """Parse the build from a dependency file."""
-        return _parse_name_version_spec_file(self.filename)
+    def parse(self, constraints_file: StrPath | None = None) -> DependencyFileData:
+        """
+        Parse the build from a dependency file.
+
+        :arg constraints_file: Extra constraints file to apply to the
+            DependencyFileData object
+        """
+        data = _parse_name_version_spec_file(self.filename)
+        if not constraints_file:
+            return data
+        constraints = parse_pieces_mapping_file(constraints_file)
+        for collection, constraint in constraints.items():
+            if collection.startswith("_"):
+                raise InvalidFileFormat(f"Invalid key: {collection}", constraints_file)
+
+            if collection not in data.deps:
+                raise InvalidFileFormat(
+                    f"{collection} is pinned in the constraints file"
+                    f" but not listed in the build file ({self.filename})",
+                    constraints_file,
+                )
+            data.deps[collection] = constraint
```

Comment:
**Replacing** the range with a constraint is IMO not what a constraint should be. Assume you have the range `>= 1.0.0, < 2.0.0` and add the constraint `< 1.5.0`. Then you end up with `< 1.5.0`, which is wrong, it should be `>= 1.0.0, < 1.5.0`. Modifying `deps.data[collection]` to `f"{deps.data[collection],{constraint}"` is generally better. On the other hand, the feature freeze feature will then complain about the result.

I don't see a way around parsing both ranges to properly combine them here, if you want to combine them in this piece of code.

Reply:
Even worse would be a constraint of `!= 1.5.0`. Then suddenly `2.0.0` could be chosen, even though the original range was `>= 1.0.0, < 2.0.0`.

Reply:
Would it be better if we named it an overrides file instead of a constraints file or are you entirely opposed to this type of solution?

Reply:
Hmm, I guess that would work too, though I personally find constraint files more useful than override files.

Maybe the whole merging shouldn't be done as part of `BuildFile.parse()`, but separately, so this can be implemented more flexible without having to make `BuildFile.parse()` know the difference between constraint and override files?